docs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.md

```diff
@@ -1,3 +1,570 @@
+        valueFrom:
+          # Use the url field of the httpProxy variable.
+          variable: httpProxy.url
+      - op: add
+        path: /spec/template/spec/dnsServers
+        valueFrom:
+          # Use the entire dnsServers array.
+          variable: dnsServers
+      - op: add
+        path: /spec/template/spec/dnsServer
+        valueFrom:
+          # Use the first item of the dnsServers array.
+          variable: dnsServers[0]
+```
+
+### Optional patches
+
+Patches can also be conditionally enabled. This can be done by configuring a Go template via `enabledIf`. 
+The patch is then only applied if the Go template evaluates to `true`. In the following example the `httpProxy` 
+patch is only applied if the `httpProxy` variable is set (and not empty).
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: ClusterClass
+metadata:
+  name: docker-clusterclass-v0.1.0
+spec:
+  ...
+  variables:
+  - name: httpProxy
+    schema:
+      openAPIV3Schema:
+        type: string
+  patches:
+  - name: httpProxy
+    enabledIf: "{{ if .httpProxy }}true{{end}}"
+    definitions:
+    ...  
+```
+
+**Tips & Tricks**:
+
+Hard-coded values can be used to test the impact of a patch during development, gradually roll out patches, etc. .
+```yaml
+    enabledIf: false
+```
+
+A boolean variable can be used to enable/disable a patch (or "feature"). This can have opt-in or opt-out behavior
+depending on the default value of the variable.
+```yaml
+    enabledIf: "{{ .httpProxyEnabled }}"
+```
+
+Of course the same is possible by adding a boolean variable to a configuration object.
+```yaml
+    enabledIf: "{{ .httpProxy.enabled }}"
+```
+
+Builtin variables can be leveraged to apply a patch only for a specific Kubernetes version.
+```yaml
+    enabledIf: "{{ if eq "v1.21.1" .builtin.controlPlane.version }}true{{end}}"
+```
+
+### MachineDeployment variable overrides
+
+If you want to use many variations of MachineDeployments in Clusters, you can either define 
+a MachineDeployment class for every variation or you can define patches and variables to 
+make a single MachineDeployment class more flexible.
+
+In the following example we make the `instanceType` of a `AWSMachineTemplate` customizable.
+First we define the `workerMachineType` variable and the corresponding patch:
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: ClusterClass
+metadata:
+  name: aws-clusterclass-v0.1.0
+spec:
+  ...
+  variables:
+  - name: workerMachineType
+    required: true
+    schema:
+      openAPIV3Schema:
+        type: string
+        default: t3.large
+  patches:
+  - name: workerMachineType
+    definitions:
+    - selector:
+        apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+        kind: AWSMachineTemplate
+        matchResources:
+          machineDeploymentClass:
+            names:
+            - default-worker
+      jsonPatches:
+      - op: replace
+        path: /spec/template/spec/instanceType
```

Comment:
I think including the templates in these yamls would be great to show exactly where the patch is happening.

Reply:
Good point here, as it helps to illustrate the example and it doesn't add to much YAML. Done.